File: app/src/main/java/com/example/notes/presentation/edit/EditFragment.java
//package com.example.notes.presentation.edit;
//
//import android.os.Bundle;
//import android.view.View;
//
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//import androidx.fragment.app.Fragment;
//import androidx.lifecycle.Observer;
//import androidx.lifecycle.ViewModelProviders;
//import androidx.navigation.Navigation;
//
//import com.example.noteEntityKotlins.R;
//
//import java.util.List;
//
//public class EditFragment extends Fragment {
//
//
//    private EditViewModel mEditViewModel;
//    private EditView mEditView;
//
//    public EditFragment() {
//        super(R.layout.fragment_edit);
//    }
//
//    @Override
//    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
//        super.onViewCreated(view, savedInstanceState);
//        mEditViewModel = ViewModelProviders.of(this).get(EditViewModel.class);
//        mEditView = new EditView(view, new EditView.Callback() {
//
//            @Override
//            public void onSaveClick(Note noteEntityKotlin) {
//                // TODO: 24/05/2021
//
//                mEditViewModel.updateNote(noteEntityKotlin);
//                Navigation.findNavController(view).popBackStack();
//
//            }
//
//            @Override
//            public void onBackPressed() {
//                Navigation.findNavController(view).popBackStack();
//            }
//        });
//        getBundle();
//
//
//        mEditViewModel.getGroupNameFromNote().observe(getViewLifecycleOwner(), new Observer<List<String>>() {
//            @Override
//            public void onChanged(List<String> list) {
//                mEditView.setListAndCreateAdapter(list);
//            }
//        });
//    }
//
//
//    private void getBundle() {
//        if (getArguments() != null) {
//            int id = getArguments().getInt("id");
//            mEditView.setNote(mEditViewModel.getNote(id));
//        }
//
//    }
//}

File: app/src/main/java/com/example/notes/features/edit/ui/EditFragment.kt
package com.example.notes.features.edit.ui

import android.content.Context
import android.os.Bundle
import android.view.*
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.fragment.findNavController
import com.example.notes.R
import com.example.notes.databinding.FragmentEditBinding
import com.example.notes.di.Injector
import com.example.notes.di.viewModel.ViewModelFactory
import com.example.notes.features.edit.commands.OnSaveEditCommand
import com.example.notes.features.edit.presentation.EditViewModel
import com.example.notes.features.notes.ui.NotesFragment
import com.example.notes.features.notes.ui.OnSetSupportActionBarCallback
import com.example.notes.models.NoteUi
import com.example.notes.utility.executeCommand
import javax.inject.Inject
import kotlin.properties.Delegates


class EditFragment : Fragment(R.layout.fragment_edit) {

    private lateinit var viewBinding: FragmentEditBinding
    private var noteId by Delegates.notNull<Int>()
    private lateinit var noteUi: NoteUi

    @Inject
    lateinit var factoryEditViewModel: ViewModelFactory
    val viewModel by viewModels<EditViewModel> { factoryEditViewModel }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        noteId = requireArguments().getInt(NotesFragment.NOTE_ID)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        Injector.inject(this)
        viewBinding = FragmentEditBinding.inflate(layoutInflater)
        return viewBinding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        initData()
        initListeners()
        initObservers()
    }

    private fun initListeners() {
        viewBinding.saveEditButton.setOnClickListener {
            executeCommand(
                OnSaveEditCommand(
                    findNavController(),
                    viewBinding,
                    viewModel
                )
            )
        }
    }

    private fun initData() {
        viewModel.getNoteById(noteId)
        viewModel.getAllNameOfGroups()
    }

    private fun initObservers() {
        viewModel.noteUi.observe(viewLifecycleOwner) { note ->
            noteUi = note
        }

        viewModel.allNameOfGroups.observe(viewLifecycleOwner) {
            setListInAutoCompleteText(it.toList())
        }
    }

    private fun setListInAutoCompleteText(list: List<String>) {
        val mArrayAdapter =
            ArrayAdapter(requireContext(), R.layout.support_simple_spinner_dropdown_item, list);
        viewBinding.autoCompleteTextView.setAdapter(mArrayAdapter);

        list.find { noteUi.nameGroup == it }?.let {
            viewBinding.autoCompleteTextView.setText(
                mArrayAdapter.getItem(mArrayAdapter.getPosition(it)), false
            )
        }
    }

    override fun onStop() {
        super.onStop()
        Injector.clearEditComponent()
    }

    companion object {
        fun newInstance() = EditFragment()
    }
}
